File: Fortran90/LinearRegression.f90
!Programmer:Kristoffer Larson
!Date: February 2, 2014
!
!Description: This was originally given in one of my classes
!  and I decided to adjust it so that it would run as a 
!  Fortran90 program. There are some adjustments that need 
!  to be made before it's fully functional.


      DIMENSION X(52), Y(2,50), LITERL(2)
      DOUBLE PRECISION S1,S2,S3,S4,S5,T,S,B,D,R,E1,E2,BBAR
      WRITE (*,10)
10    FORMAT('0',1X,'*   *   *  LINEAR REGRESSION ANALYSIS  *   *   *',//)
      WRITE (*,20)
20    FORMAT(1X,'HOW MANY PAIRS TO BE ANALYZED?'$)
      READ (5,*) N
      IF (N.GT.50) GOTO 70
      WRITE (*,30)
30    FORMAT(//1X,'Enter one pair at a time')
      WRITE (*,40)
40    FORMAT(1X,'and separate X from Y with a comma.'//)
      WRITE (*,50)
50    FORMAT(1X,'Enter pair number one : '$)
      READ (5,*) X(1), Y(1,1)
         DO 60 I=2,N
         WRITE (*,55) I
55       FORMAT(1X,'Enter pair number',I3,' : '$)
         READ (5,*) X(I), Y(1,I)
60       CONTINUE
      GOTO 90
70    WRITE (*,80)
80    FORMAT(1X,'At present this program can only handle 50 data pairs.')
      STOP
90    WRITE (*,100)
100   FORMAT(1X,'Would you like to review the data?'$)
      READ (5,110) LITERL(1)
110   FORMAT(A1)
      IF (LITERL(1) == 78) GOTO 140
         DO 130 I=1,N
         WRITE (*,120) I, X(I), Y(1,I)
120      FORMAT(1X,'DATA PAIR',I3,').  ',2F20.10)
130      CONTINUE
140   WRITE (*,150)
150   FORMAT(1X,'Would you like to make any changes?'$)
      READ (5,110) LITERL(1)
      !IF (LITERL(1) == 78) GOTO 200
      WRITE (*,160)
160   FORMAT(1X,'To change a data pair, enter the')
      WRITE (*,165)
165   FORMAT(1X,'Pair Number and the new X, Y pair.')
      WRITE (*,180)
180   FORMAT('0',1X,'How many changes shall we make?'$)
      READ (5,*) ICHANG
         DO 190 I=1, ICHANG
         WRITE (*,185)
185      FORMAT(1X,'CHANGE NUMBER',I4,' : '$)
         READ (5,*) C1, C2, C3
!         X(C1)=C2
!         Y(1,C1)=C3
190      CONTINUE
      GOTO 90
200      DO 210 I=1,N
         S1=S1+X(I)
         S2=S2+Y(1,I)
         S3=S3+X(I)*Y(1,I)
         S4=S4+X(I)*X(I)
         S5=S5+Y(1,I)+Y(1,I)
210      CONTINUE
      T=N*S4-S1*S1
      S=(N*S3-S1*S2)/T
      B=(S4*S2-S1*S2)/T
         DO 220 I=1,N
         Y(2,I)=S*X(I)+B
         D=D+(Y(2,I)-Y(1,I))**2
220      CONTINUE
      D=D/(N-2)
      E1=DSQRT(D*N/T)
      E2=DSQRT(D/N*(1+S1*S1/T))
      R=(N*S3-S1*S2)/(DSRT(ABS(((N*S4-ABS(S1)**2))*(N*S5-ABS(S2)**2))))
      WRITE (*,230)
230   FORMAT(////,10X,'X-VALUE',20X'Y-OBS',22X,'Y-CALC')
      WRITE (*,235)
235   FORMAT(10X,'=======',20X,'=====',22X,'======')
         DO 250 I=1,N
         WRITE (*,240) X(I), Y(1,I), Y(2,I)
240      FORMAT(F22.10,2F27.10)
250      CONTINUE
      WRITE (*,260) S
260   FORMAT(//,1X,'SLOPE = ',D22.16)
      WRITE (*,270) E1
270   FORMAT(1X,'THE ERROR IN THE SLOPE IS +/- ',D22.16)
      WRITE (*,280) B
280   FORMAT(//,1X,'INTERCEPT = ',D22.16)
      WRITE (*,285) E2
285   FORMAT(1X,'THE ERROR IN THE INTERCEPT IS +/- ',D22.16)
      !LITERL(2)='+'
      !IF (ABS(B).NE.B) LITERL(2)='-'
      WRITE (*,290)
290   FORMAT(//,1X,'EQUATION FOR THE BEST LINEAR FIT IS : ')
      BBAR=ABS(B)
      WRITE (*,295) S,LITERL(2),BBAR
295   FORMAT(1X,'Y(X) =',D25.16,' * X ',A1,D25.16,///)
      WRITE (*,296) R
296   FORMAT(' LINEAR CORRELATION COEFFICIENT =',D22.16)
      STOP 'LINEAR... Execution completed'
      End

File: Java/LinearRegression.java
//    Programmers: Kristoffer Larson
//    Date: February 19, 2014
//    
//    Description: A translation of a fortran linear regression program.
//
//
//
//    Irvin Levy (Gordon College) Linear Regression Package
//    
//    *  *  *  *  VARIABLE LIST  *  *  *  *
//             S1 - Sum of X values
//             S2 - Sum of Y values
//             S3 - Sum of X * Y
//             S4 - Sum of X * X
//             S5 - Sum of Y * Y
//             D  - Sum of Y(obs) - Y(calc) squared
//             S  - Slope
//             B  - Y-intercept
//             E1 - Error in the slope
//             E2 - Error in the intercept
//             T  - Beta function for linear regression
//             R  - Correlation coefficient
//    
//    Error function uses (N-2) weigthing
// 
// 
// 
import java.util.Scanner;

public class LinearRegression {
   public static void main(String[] args) {
      double[] x = new double[52]; double[][] y = new double[2][50]; String in, sign;
      double S1 = 0, S2 = 0, S3 = 0, S4 = 0, S5 = 0, T = 0, S = 0; 
      double B = 0, D = 0, R = 0, E1 = 0, E2 = 0, BBAR = 0;
      
      Scanner input = new Scanner(System.in);
      System.out.printf("0 *   *   *   LINEAR REGRESSION ANALYSIS   *   *   *\n\n");
      System.out.printf(" HOW MANY PAIRS TO BE ANALYZED?");
      int n = input.nextInt();
      input.nextLine();
      
      if (n > 50) {
      System.out.printf(" At present this program can only handle 50 data pairs.");
      System.exit(1);
      }
      
      System.out.printf("\n\n Enter one pair at a time");
      System.out.printf("\n and separate X from Y with a comma.\n\n");
      System.out.printf(" Enter pair number one : ");
      String point = input.nextLine();
      String[] splits = point.split(",");
      x[0] = Double.parseDouble(splits[0]);
      y[0][0] = Double.parseDouble(splits[1]);
      
      for (int i = 1; i < n; i++) {
         System.out.printf(" Enter pair number %d : ", i + 1); 
         point = input.nextLine();
         splits = point.split(",");
         x[i] = Double.parseDouble(splits[0]);
         y[0][i] = Double.parseDouble(splits[1]);
      }
      
      boolean test = true;
      while (test) {
         System.out.printf(" Would you like to review the data?");
         in = input.next();
         if (!(in.compareTo("N") == 0)) {
            for (int i = 0; i < n; i++) {
            System.out.printf(" DATA PAIR %d).%20.10f%20.10f\n", i + 1, x[i],y[0][i]);
            }
         }
         System.out.printf(" Would you like to make any changes?");
         in = input.next();
         if (!(in.compareTo("N") == 0)) {
            System.out.printf(" To change a data pair, enter the\n");
            System.out.printf(" Pair Number and the new X, Y pair.\n");
            System.out.printf("0 How many changes shall we make?");
            int iChang = input.nextInt();
            input.nextLine();
            int change;
            for (int i = 0; i < iChang; i++) {
               System.out.printf("CHANGE NUMBER\n");
               change = input.nextInt();
               point = input.nextLine();
               splits = point.split(",");
               x[change - 1] = Double.parseDouble(splits[0]);
               y[0][change - 1] = Double.parseDouble(splits[1]);
            }
         } else 
            test = false;
      }
      for (int i = 0; i < n; i++) {
         S1 = S1 + x[i];
         S2 = S2 + y[0][i];
         S3 = S3 + x[i] * y[0][i];
         S4 = S4 + x[i] * x[i];
         S5 = S5 + y[0][i] * y[0][i];
      }
      T = n * S4 - S1 * S1;
      S = (n * S3 - S1 * S2) / T;
      B = (S4 * S2 - S1 * S3) / T;
      for (int i = 0; i < n; i++) {
         y[1][i] = S * x[i] + B;
         D = D + Math.pow((y[1][i] - y[0][i]), 2);
      }
      D = D / (n - 2);
      E1 = Math.sqrt(D * n / T);
      E2 = Math.sqrt(D / n * (1 + S1 * S1 / T));
      double R1 = (n * S4 - Math.pow(Math.abs(S1), 2));
      double R2 = (n * S5 - Math.pow(Math.abs(S2), 2));
      R = (n * S3 - S1 * S2) / (Math.sqrt(Math.abs(R1 * R2)));
      
      System.out.printf("\n\n\n\n%10s%20s%22s\n","X-VALUE","Y-OBS","Y-CALC");
      System.out.printf("%10s%20s%22s\n", "=======", "=====", "======");
      
      for (int i = 0; i < n; i++) {
         System.out.printf("%15.10f %21.10f %20.10f\n", x[i], y[0][i], y[1][i]);
      }
      
      System.out.printf("\n\n SLOPE = %22.16f\n", S);
      System.out.printf(" THE ERROR IN THE SLOPE IS +/- %22.16f\n", E1);
      System.out.printf("\n\n INTERCEPT = %22.16f\n", B);
      System.out.printf(" THE ERROR IN THE INTERCEPT IS +/- %22.16f\n", E2);
      
      sign = "+";
      if(Math.abs(B) != B) sign = "-";
      
      System.out.printf("\n\n EQUATION FOR THE BEST LINEAR FIT IS : \n");
      BBAR = Math.abs(B);
      System.out.printf(" Y(X) = %25.16f * X %s %25.16f\n\n\n", S, sign, BBAR);
      System.out.printf(" LINEAR CORRELATION COEFFICIENT = %22.16f\n", R);
      System.out.print("LINEAR... Execution completed");
   }
}
